### ISSUE TITLE
Bug, accessibility, front-end: Keyboard hotkeys depend on the keyboard layout

### ISSUE THREAD

**Issue by inga-lovinde** [tootsuite/mastodon#15348]
### Expected behaviour

Hotkeys should work as described in "Hotkeys" table. 

### Actual behaviour

Hotkeys are handled according to the letter entered, not according to the key pressed.

### Steps to reproduce the problem

* Switch to Dvorak layout.

    * In timeline, select the toot  and press "X" key. Instead of expanding the toot (CW), it does nothing (as if "Q" was pressed, which is the char produced by Dvorak layout when pressing on "X" key).

    * Press "R" key. Instead of replying to the toot, it opens the author's profile (as if "P" was pressed, which is the char produced by Dvorak layout when pressing "R" key).

    * etc...

* Switch to non-Latin (e.g. Russian) layout

    * In timeline, select the toot and press "X" key. Nothing happens.

    * In timeline, select the toot and press "R" key. Nothing happens.

    * etc...

Note that according to https://developer.mozilla.org/en-US/docs/Web/API/GlobalEventHandlers/onkeydown , `onkeydown` event receives the actual key pressed, no matter what layout is selected (it's "KeyX" when I press key "X" regardless whether it's Dvorak layout which produces "Q" or standard Russian layout which produces "Ч"). So `onkeydown` works correctly, and the bug is somewhere in Mastodon.

### Specifications

Mastodon v3.2.1, Firefox 83.0, Windows 10 20H2.

**Comment by Gargron** [comment]
This is consistent with how keybinds work in other applications in my experience, i.e. when I switch from US to German layout, my "z" keybind in Dota 2 moves from the bottom left to the top middle. You must consider that the user must be able to find the hotkey on their keyboard. If the "z" hotkey were hardcoded to be on the bottom left, a German user would not be able to find it, since they would be trying to press the button on their keyboard that is labelled "z" but which is in a different place. So this is not a bug. I do not have a solution for non-ascii keyboard layouts.

**Comment by Gargron** [comment]
(Theoretically every non-ascii locale would need to have its own set of hotkeys defined, but it's not possible for me to do since I am not a speaker of every language and do not know their keyboard layouts, also, selected UI locale may not always correspond to keyboard layout or input language, so, this is a wontfix)

**Comment by inga-lovinde** [comment]
> This is consistent with how keybinds work in other applications in my experience

This is absolutely inconsistent with how hotkeys work in other applications in _my_ experience, and I would say that probably the apps you refer to are also buggy?

A few examples off the top of my mind:

* Win+L always locks the screen on Windows. I can just press these two keys on every PC not thinking what layout is, and it will lock. There is no such thing as "press Win+P if you're using dvorak (because in dvorak "P" key produces "L" letter)", there is no such thing as "if you're now in Russian layout, switch to English layout in order to lock the screen".

* Ctrl+S always saves the document in Notepad. I don't have to think what layout is currently active, again.

* Ctrl+X, Ctrl+C, Ctrl+V always cut/copy/paste. These three combos are next to each other for a reason, and merely switching to dvorak does not mean they are now in a different corners of a keyboard. Switching to Russian layout does not mean I can no longer cut/copy/paste using hotkeys.

* WASD, obviously. In every game I've used, WASD continues to work as WASD, these keys also do not stop working when I'm switching for Russian layout, and they also don't turn into "," for forward, "A" for left, ";" for backward, "H" for right - that would make no sense and would make these hotkeys totally unusable!

Keyboard layout defines how the text is entered, it should not define how hotkeys are working, they are called "hotkeys", not "hot letters". And "onkeydown" event produces the _key_ pressed instead of letter probably for the same reason.

> i.e. when I switch from US to German layout, my "z" keybind in Dota 2 moves from the bottom left to the top middle.

If you're on Windows: open Notepad, enter some text, Ctrl+Z to undo.

Or in browser input area: enter some text, Ctrl+Z to undo, Ctrl+Y to redo. Undo and redo do not magically change their places when I switch between US English and German layouts - at least on Firefox 83 on Windows 10 20H2 and every other browser I tried.

> (Theoretically every non-ascii locale would need to have its own set of hotkeys defined, but it's not possible for me to do since I am not a speaker of every language and do not know their keyboard layouts, also, selected UI locale may not always correspond to keyboard layout or input language, so, this is a wontfix)

If I understand you correctly, you are saying that hotkeys won't work in non-Latin locales (and that for people who use non-English locales it will depend on what layout is currently active - so that hotkeys will only work half the time) because it will be impossibly hard to make them work while preserving that "changing between different Latin layouts changes all the hotkeys".

That sounds to me like "Mastodon hotkeys will remain inaccessible to non-Latin locales because then we would have to make them accessible to non-QWERTY Latin locales", and honestly, I don't see the problem here. Just make them accessible to everybody - look at the key pressed and not at the character entered.

Please, reopen this bug.

**Comment by inga-lovinde** [comment]
Also consider this case: a bilingual person is using standard German layout to enter German text and Dvorak to enter English text. So all their Mastodon hotkeys now "work" but react differently depending on the current layout. For example, a user wants to "focus the compose textarea", presses "N" key, and it works if they were entering German text before; but if they were entering English text before, the same key boosts the current toot instead.

That means that before using any hotkey, user has to check their current layout, or make sure the correct one is selected (one for which they memorized the hotkeys location), every time they want to use a hotkey, and this is kind of defeats the entire purpose of hotkeys.

**Comment by codl** [comment]
AZERTY user of 15 ish years here

"Shortcuts move when you change your keyboard layout" has been my experience everywhere except a few video games

I would be surprised and annoyed if a program decided that, actually, ctrl+z is ctrl+w because that's what would be printed on that key on a QWERTY keyboard

**Comment by codl** [comment]
I wonder if capturing both keycode and scancode, and using the former if it's non-latin, the latter mapped to QWERTY otherwise, would at least make the shortcuts available to non-latin layouts... assuming that there are no layouts that are half-latin and half-some-other-script.

**Comment by inga-lovinde** [comment]
> I would be surprised and annoyed if a program decided that, actually, ctrl+z is ctrl+w because that's what would be printed on that key on a QWERTY keyboard

That's really odd. What happens if you press Ctrl + the button in the bottom left corner while in toot compose area or in comment compose area on GitHub? For me, this undoes the latest action (e.g. the latest piece of text entered), regardless of whether it's QWERTY or QWERTZ or Dvorak or standard Russian layout, in all browsers on Windows 7 and Windows 10 at least.

**Comment by codl** [comment]
> What happens if you press Ctrl + the button in the bottom left corner while in toot compose area or in comment compose area on GitHub?

If I'm in AZERTY, the tab closes (because that's ctrl+w). If I'm in QWERTY, then it undoes the latest edit as you described.

I'm on linux at the moment but I tried in a Windows 10 virtual machine and got the same behaviour. For that matter I wasn't able to reproduce what you described in any of the layouts you mentioned (except qwerty and russian)? I wonder if it depends on system locale or something.

**Comment by Gargron** [comment]
> > i.e. when I switch from US to German layout, my "z" keybind in Dota 2 moves from the bottom left to the top middle.
> 
> If you're on Windows: open Notepad, enter some text, Ctrl+Z to undo.
> 
> Or in browser input area: enter some text, Ctrl+Z to undo, Ctrl+Y to redo. Undo and redo do not magically change their places when I switch between US English and German layouts - at least on Firefox 83 on Windows 10 20H2 and every other browser I tried.

When I switch layouts that's literally what happens in Sublime Text 2 (haven't tried Notepad), Ctrl + Z and Ctrl + Y switch places.

I still don't quite understand how you'd be able to identify where "Ctrl + Z" is on keyboards that have a different position and number of keys if you weren't relying on the OS input interpretation.

**Comment by inga-lovinde** [comment]
Turns out this topic is deeper than I originally thought.

@codl 

> If I'm in AZERTY, the tab closes (because that's ctrl+w). If I'm in QWERTY, then it undoes the latest edit as you described.

So if you're not sure what layout is currently selected, your attempt to undo the latest edit could close the current tab instead? Sounds like really bad UX to me...

> When I switch layouts that's literally what happens in Sublime Text 2 (haven't tried Notepad), Ctrl + Z and Ctrl + Y switch places.

That's odd because almost every single app I've used over the last 7 years (or 20 years if we also count QWERTY vs. Russian layout) respected actual key presses instead of letters produced by the current layout. I can only remember a couple of programs which did not respect this, and it was so frustrating every time that it basically made hotkeys feature of these apps unusable.

> I still don't quite understand how you'd be able to identify where "Ctrl + Z" is on keyboards that have a different position and number of keys if you weren't relying on the OS input interpretation.

All keyboards (most of them) have roughly the same positions and number of keys which are then mapped to letters depending on layout.

It did not occur to me because all my keyboards were either Russian (and Russian keyboards have two letters printed on every key, one russian from ЙЦУКЕН layout and another english from QWERTY layout).

So I think I conflated two meanings here. When I say, for example, "WASD", I mean "that block of four keys that resembles inverted T on some keyboards, which is under 123 and to the right of caps lock on some keyboards". Of course on esoteric keyboards it will look entirely different; and even on plain keyboards which have "AZERTY" labels printed on them (such as ordinary keyboards sold in France) the same triangle will be labeled "ZQSD".

There are three meanings the key has:
* What is its location (it's the same key as W on US international keyboards, it's directly under 2 on some keyboards, or one key away from Tab on some keyboards), and what scancode it produces (it's "W" unless power user overrided scancodes handling in their OS)
* How it is labeled (it's labeled "W" on US international keyboards, "Z" on French keyboards, "W/Ц" on Russian keyboards, or blank on these fancy keyboards without labels);
* What letter does it produce in a text input field if pressed (it's "W" in US English, "<" in Dvorak, "Z" in standard French, "Ц" in standard Russian, etc).

For me it was self-obvious that hotkeys should work according to the first meaning, but turns out other people have other expectations. Unfortunately I did not find any literature on this topic - probably there already are some standards, but I did not find them.

You are right in the sense that a user in France who has a French-labeled keyboard and only uses French will think that the key under 2 is "Z". And if we're going to handle Ctrl+W (as in Ctrl+that key under 2) hotkey, it is not clear how to communicate to the user how do they trigger this hotkey, because for them, W is an entirely diffferent key which is next to the left long shift (or one key away from the left short shift). So we could say that hotkeys should work according to however physical keys are labeled. Since we have no way of knowing this, we could use current keyboard layout as the proxy for key labels; for a significant part of users (those who only use a single keyboard layout which matches their keyboard and uses Latin script - e.g. French users of French laptops) that would work.

On the other hand, some hotkeys are there because their location has some meaning. E.g. WASD does not have any mnemonic meaning, but it should be an inverted T. I don't think a French gamer using French laptop with AZERTY labels who only uses AZERTY layout would expect to control a game with keys which are labeled WASD on their keyboard.

And the third thing is that there could be an expectation that, no matter how hotkeys are defined, they should not change their meaning or stop working if user switches to another layout - because after some time they are wired into the muscle memory and are used automatically without thinking. If I have that muscle memory "to revert the latest change of the text, press Ctrl and this key on the bottom left", then I'm most definitely do not expect that if it's French text I'm editing, I will instead close the tab and lose the entire text; and I do not expect that if it's Russian text I'm editing, it won't work at all. That could be a source of endless frustration making the entire hotkey feature worse than useless.

It seems that the first and the third point could not be reconciled. So maybe turn this into a feature request for an additional toggle in options?

* By default, use current "char entered" method to detect hotkeys (which will work for users who only use the layout that matches their keyboard labels);
* But also allow user to switch to the "key pressed" method (which will work for users who use some other layouts or who have their keyboards labeled with QWERTY or not labeled at all or who just have memorized QWERTY).

Also note that it seems that most, if not all, keyboards sold in countries with non-Latin scripts have an auxiliary QWERTY labeling. So while for a French user it would be difficult to understand to which key "W" refers, it would be obvious for a user from Russia, China, Japan, Thailand etc; they have another problem now - that hotkeys do not work at all in their home layout.
